File: qubgrademe-meanmark/src/services/Validator.cs
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using Assignment2.MeanMark.Types;

[assembly: InternalsVisibleTo("tests")]
namespace Assignment2.MeanMark.Services;

public  class Validator : IValidator
{
    public ValidationResponse Validate(List<string> modules, List<string> marks)
    {

        var response = new ValidationResponse() { error = false, errorMessage = "" };
        // Checking to see if all modules are empty
        if (modules.Count() == modules.Count(x => !IsStringValidModule(x)))
        {
            response.error = true;
            response.errorMessage = "Please provide some module codes and their respective marks";
        }

        if (!response.error)
        {
            for (int i = 0; i < modules.Count(); i++)
            {
                // Check if valid module at this index
                if (IsStringValidModule(modules[i]))
                {
                    // Check if mark is invalid
                    if (!IsMarkValidInteger(marks[i]))
                    {
                       response.error = true;
                       response.errorMessage = "Please provide a valid integer for every entered module";
                       break;
                    }
                }
                else
                {
                    if (IsMarkValidInteger(marks[i]))
                    {
                        response.error = true;
                        response.errorMessage = "Please provide a module name for all marks entered";
                        break;
                    }
                }
                
                // If not, see if there is a valid mark
            }
        }


        return response;
    }


    private bool IsMarkValidInteger(string mark)
    {
        int intMark;
        if (int.TryParse(mark, out intMark))
        {
            return (intMark <= 100 && intMark >= 0);
        }

        return false;
    }

    private bool IsStringValidModule(string module) => !string.IsNullOrWhiteSpace(module);

}

File: qubgrademe-classifygrade/src/main/java/validation/Validator.java
package validation;

import java.util.ArrayList;
import java.util.List;

public class Validator {
    public static String validate(List<String> modules, List<String> marks){

        if(allEmpty(modules)){
            return "Please provide some module codes and their respective marks";
        }
        for(int i = 0; i < modules.size(); i++){
            // Check if there is a valid mark at this index
            if(modules.get(i).trim() != ""){
                // Check if there is not a valid integer at this point
                if(!validateInt(marks.get(i)))
                    return "Please provide a valid integer for every entered module";
            }else{
                // If there isnt a valid mark
                // Check if there is a valid integer
                if(validateInt(marks.get(i)))
                    return "Please provide a module name for all marks entered";
            }
        }
        return "";
    }

    public static boolean allEmpty(List<String> modules){
        return modules.size() == modules.stream().filter(m -> m.trim() == "").toList().size();
    }

    public static boolean validateInt(String mark){
        try{
            int num = Integer.parseInt(mark);
            if(num >= 0 && num <= 100)
                return true;
        }catch(NumberFormatException ignored){}
        return false;
    }

    public static List<Integer> marksToInteger(List<String> marks){
        return marks.stream().map(m -> validateInt(m) ? Integer.parseInt(m) : 0).toList();
    }
}
